grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -50,14 +65,32 @@ def _GetTruncatedTable(
   return result
 
 
-class OsqueryFlow(flow_base.FlowBase):
+def _ExtractFileInfo(
+    result: rdf_flow_objects.FlowResult,
+) -> rdf_osquery.OsqueryCollectedFile:
+  if not isinstance(result.payload, rdf_client_fs.StatEntry):
+    raise _ResultNotRelevantError
```

Comment:
I would just throw `TypeError` and not bother with additional exception type.

Reply:
Isn't the code more readable if the exception has a descriptive name at the place it is handled? https://github.com/google/grr/pull/901/files/31944c1ea3aadc3c54dc6e2f459464439685ef8a#diff-ad469360f11bc0231fc5350d52ad08e36d21fb93828c0f47aa45fa24393b54bbR242

Reply:
I didn't realize you catch it, you do then it makes sense. But then I am not sure if it is good behaviour: if there is a type error it means something is seriously wrong. If we silently ignore such errors, the system will pretend everything is fine even though it is not.

Reply:
It's not really type **error**, it really is that the result is not relevant. Maybe I should remove the "Error" in _ResultNotRelevantError"? Our flow sends two types of results: `OsqueryResult` for the actual osquery table results, and `StatEntry` for each collected file. This is because of `GetFilesArchiveMappings`: it is called with the results generated by the flow (in order to generate the file archive).

Reply:
> Our flow sends two types of results: OsqueryResult for the actual osquery table results, and StatEntry for each collected file.

This is very unfortunate, actually. Flows should be well typed and return a specific type. Wouldn't it be possible to split `OsqueryResult` into `OsqueryResult` (for client action) and `OsqueryFlowResult` (for the flow), similarly how you do with the args, and add a `collected_files` field into the flow one?

That would make it correctly typed, avoids having this awkward `isinstance` logic and does not abuse exceptions for control flow.

Reply:
Which one of the two are you suggesting (or maybe something else?):
1. Bundling osquery results with flow collection results.
2. Making `OsqueryFlowResult` effectively a union type of the current `OsqueryResult` and `StatEntry`, and just populating it with one or the other in the result callback methods (`Process` and `ReceiveFetchedFile`)?

Also by "and does not abuse exceptions for control flow" do you mean that the code is difficult to read in its current version, or maybe that it is inefficient? From my understanding it is not uncommon to use exception for control flow in Python, since they do not add much overhead. Also the construction looks at least as clear as having an if clause.

Reply:
The first option. I think it makes sense to have a field with stat entries for the collected file, it is part of the result afterall.

> From my understanding it is not uncommon to use exception for control flow in Python

It is not uncommon, but I think it is unfortunate. Afterall, as the name suggests, exceptions are for exceptional situations, which we clearly not have here. It's not that they incur some performance hit (I mean it is Python, we don't care about that) but rather make the code feel awkward: you raise but with the assumption that the caller is going to catch this—however this cannot be specified in any way at the type level.

> Also the construction looks at least as clear as having an if clause.

For now, because you just wrote it and you know all the assumptions. Once the code evolves and is read by others, it might no longer be the case. I am not saying this should never be done (I did it myself couple times!), but if there is an easy solution that does not involve using this mechanism, why not?

Reply:
> The first option. I think it makes sense to have a field with stat entries for the collected file, it is part of the result afterall.'

I like the idea, but I'm not sure it is suitable for this case, as it would introduce the following complications:

1. We receive the two kinds of results at different places: osquery results in the `Process` callback, and file collection results in the `ReceiveFetchedFile` callback. Thus we would need to add some state and logic in order to bundle them together.
2. Osquery results are received and inspected in full before file collection is initiated. In order to bundle them with the file collection results, osquery results would need to sit in memory while the file collection stage completes. I guess this wouldn't be a big issue now, since we are keeping all osquery results in memory anyways (as opposed to iterating through them), but it would stop us from optimizing this in the future too.
3. We send requests for file collection one by one, and receive the results one by one. We would need some mechanism to determine when we have received all files we have requested (or errors for them). And if for some reason we don't receive a response for all files we attempted to collect, then our flow will hang forever, waiting for them, so that they can be bundled and sent together with the osquery results (that wait in memory).

What are your thoughts on those issues?

Reply:
> Thus we would need to add some state and logic in order to bundle them together.

It is extremely trivial logic (put a value into the flow state, get a value from the flow state), I don't think there is anything to be worried here.

> I guess this wouldn't be a big issue now, since we are keeping all osquery results in memory anyways (as opposed to iterating through them), but it would stop us from optimizing this in the future too.

Sounds a bit like premature optimization. If the list of files to download is so large that there are problems with keeping it in memory, it is likely not a list you would ever want to fetch from the client.

> We send requests for file collection one by one, and receive the results one by one.

Why? The whole point of `MultiGetFile` is that it can do multiple files. And why do we use `MultiGetFileMixin` rather than spawning a subflow? The whole mixin system in GRR is extremely fragile and depends on method names to glue everything together, so I would prefer to stay away from this model.

Reply:
> Sounds a bit like premature optimization. If the list of files to download is so large that there are problems with keeping it in memory, it is likely not a list you would ever want to fetch from the client.

My point was that we would have to keep the osquery results in memory, not the list of files. However, I just realized that we enforce limits on the table size for file collection (at most 1000 rows), so it would not be as bad as having millions of rows in memory.

> And why do we use `MultiGetFileMixin` rather than spawning a subflow?

We would like to have full information about which files were successfully collected in full, which were truncated, and which encountered an error, so that we report it to the user. This is not currently supported by `MultiGetFile`. Unfortunately it is out of the scope of this PR to improve `MultiGetFile`, and this is why we went with `MultiGetFileLogic`.

Maybe @mbushkov has something to add to this discussion?

Reply:
Interesting discussion!

There are a few practical issues with extending OsqueryResult to include the files. The Osquery client action takes measures to not grow OsqueryResult protobufs too much. I think, to be honest, that we should store the actual osquery output in the blobstore and just keep references to blobs in the results, since having 2Mb-sized rows in the database is suboptimal, but that's another discussion.

If we introduce the OsqueryFlowResult by adding file stats to the OsqueryResult, we have to make sure that the proto doesn't grow too much. If the paths are long and we download 1000 files, how do we guarantee that it stays within a reasonable limit? I mean, it should, but it feels like we should at least do some check and checking protos sizes is awkward.

Another practical concern: you would need to fetch all the files first and only then return results and this breaks the practice of flows returning results incrementally, as soon as they arrive.

> Flows should be well typed and return a specific type. Wouldn't it be possible to split OsqueryResult into OsqueryResult (for client action) and OsqueryFlowResult (for the flow), similarly how you do with the args, and add a collected_files field into the flow one?

We do have flows that return multiple things and I think this is a valid concept, primarily because it keeps combining flows easy/trivial. If your flow runs a bunch of other flows, you can always just forward the results. This is heavily used, for example, by the ArtifactCollector. This is what was done in the OsqueryFlow initially with MultiGetFile flow. The reason why Simeon switched to MultiGetFileLogic is because MultiGetFile doesn't return collection errors, but again - this is another problem.

I would suggest to have a wider discussion on the question of returning multiple types of results, but do this outside of the scope of this PR. 

This PR largely supports an already finished UX PR that introduces the Osquery files collection functionality to the UI - https://github.com/google/grr/pull/906. Refactoring the approach used here would also require a significant refactoring of the UX PR. Given the time constraints and the concerns I outlined above and the fact that we don't have a policy of "flows return one thing" yet, I'd suggest proceeding with the current implementation and, if needed, agreeing on future refactorings. @panhania, what do you think?

Reply:
> I think, to be honest, that we should store the actual osquery output in the blobstore and just keep references to blobs in the results, since having 2Mb-sized rows in the database is suboptimal

But it is so only in some extreme cases. Usually, results of osquery queries are tiny. Cases where they are not (e.g. listing users on machines where there are hundred thousands of them) would give large results also using ordinary flows.

I don't think it makes sense to complicate logic and add overhead to accommodate for corner cases that other flows don't even consider (the fact that the osquery flow is able to handle them semi-gracefully is just a nice plus).

> If we introduce the OsqueryFlowResult by adding file stats to the OsqueryResult, we have to make sure that the proto doesn't grow too much. If the paths are long and we download 1000 files, how do we guarantee that it stays within a reasonable limit?

I would say: if it doesn't fit, it should not be collected. Maximum path length is 4096 bytes as far as I remember on Linux (on other platforms even less), any reasonable results will easily fit into proto. Other flows also don't consider extreme corner cases.

> We do have flows that return multiple things and I think this is a valid concept, primarily because it keeps combining flows easy/trivial. If your flow runs a bunch of other flows, you can always just forward the results. This is heavily used, for example, by the ArtifactCollector. This is what was done in the OsqueryFlow initially with MultiGetFile flow.

I mean, the fact that a broken mechanism is used in abusive way to implement other broken mechanisms is not a good argument. Because of how "easy/trivial" it is to combine flows with this weakly typed system is exactly the reason why nobody on the team dares to even look at the Frankenstein-style giant pile of intangible mess sewed together from several other mixins that we innocently called "the file finder flow".

The same way you could argue that having functions that return multiple types is a valid concept. And I am sure that there are examples of code that uses such travesties (even in this very codebase) but this is not the way one should program and we already learned this the hard way.

Code is not supposed to be easy or trivial to write. The only thing that matters is how easy it is to read it, reason about it and maintain it.

Reply:
We had an offline discussion covering the points above.

### `MultiGetFile` flow vs. `MultiGetFileLogic` mixin
We agreed to use the `MultiGetFileLogic` mixin instead of the `MultiGetFile` flow for the time being. Our main point was that `MultiGetFile` doesn't give precise per-file error information. However, after the discussion I realized that it actually does: it populates a list with file collection outcomes [in its state](https://github.com/simstoykov/grr/blob/filepaths_collection_backend/grr/server/grr_response_server/flows/general/transfer.py#L836), and makes it available [in its progress](https://github.com/simstoykov/grr/blob/filepaths_collection_backend/grr/server/grr_response_server/flows/general/transfer.py#L796).

There was another point, which is still valid: `MultiGetFileLogic` allows us to collect files one by one, incrementally. This is consistent with the current approach, because before we send requests for collection we first stat files (to check their size). The stat results are received one by one, so it is simpler to send requests for collection one by one again, instead of having to merge them together.

At this point the logic is not complicated or coupled with anything, so it should be easy to change `MultiGetFileLogic` mixin use to `MultiGetFile` flow use in the future.

### Bundling file collection and osquery results into one protobuf, so that the flow produces one type of results
This discussion produced what seems to be a better design than the current one: we split the joint osquery-and-file-collection flow into two flows - one for osquery, and another one for collection files from osquery results. In this way we can have a unique flow result type for each of the two, and we also don't need to buffer and bundle results together. Sadly, this is beyond the scope of this PR and my internship. That's why we decided to proceed with the current design, and implement the improvement in the future.